src/twisted/python/test/test_systemd.py

```diff
@@ -7,158 +7,163 @@
-    C{inheritedDescriptors} which is based on the environment variables set by
-    systemd.  To facilitate testing, this mixin will also create a fake
-    environment dictionary and add keys to it to make it look as if some
-    descriptors have been inherited.
+    Apply tests to L{ListenFDs}, constructed based on an environment dictionary.
     """
 
-    def initializeEnvironment(self, count, pid):
-        """
-        Create a copy of the process environment and add I{LISTEN_FDS} and
-        I{LISTEN_PID} (the environment variables set by systemd) to it.
-        """
-        result = os.environ.copy()
-        result["LISTEN_FDS"] = str(count)
-        result["LISTEN_PID"] = str(pid)
-        return result
+    @given(lists(systemd_descriptor_names(), min_size=0, max_size=10))
+    def test_fromEnvironmentEquivalence(self, names: Sequence[str]) -> None:
+        """
+        The L{ListenFDs} and L{ListenFDs.fromEnvironment} constructors are
+        equivalent for their respective representations of the same
+        information.
+
+        @param names: The names of the file descriptors to represent as
+            inherited in the test environment given to the parser.  The number
+            of descriptors represented will equal the length of this list.
+        """
+        numFDs = len(names)
+        descriptors = list(range(ListenFDs._START, ListenFDs._START + numFDs))
+        fds = ListenFDs.fromEnvironment(
+            {
+                "LISTEN_PID": str(os.getpid()),
+                "LISTEN_FDS": str(numFDs),
+                "LISTEN_FDNAMES": ":".join(names),
+            }
+        )
+        assert_that(fds, equal_to(ListenFDs(descriptors, tuple(names))))
 
-    def getDaemon(self, start, count):
+    def test_defaultEnviron(self) -> None:
         """
-        Create a new L{ListenFDs} instance, initialized with a fake environment
-        dictionary which will be set up as systemd would have set it up if
-        C{count} descriptors were being inherited.  The descriptors will also
-        start at C{start}.
+        If the process environment is not explicitly passed to
+        L{ListenFDs.fromEnvironment}, the real process environment dictionary
+        is used.
         """
-        fakeEnvironment = self.initializeEnvironment(count, os.getpid())
-        return ListenFDs.fromEnvironment(environ=fakeEnvironment, start=start)
-
-
-class MemoryOnlyTests(MemoryOnlyMixin, InheritedDescriptorsMixin, TestCase):
-    """
-    Apply tests to L{ListenFDs}, explicitly constructed with some fake file
-    descriptors.
-    """
-
-
-class EnvironmentTests(EnvironmentMixin, InheritedDescriptorsMixin, TestCase):
-    """
-    Apply tests to L{ListenFDs}, constructed based on an environment dictionary.
-    """
+        self.patch(os, "environ", initializeEnvironment(5, os.getpid()))
+        sddaemon = ListenFDs.fromEnvironment()
+        self.assertEqual(list(range(3, 3 + 5)), sddaemon.inheritedDescriptors())
```

Comment:
I now understand what this test does, but I think the the actual environent variables are important for this tests, and they are hidden in `initializeEnvironment`

Also, for these tests, I prefer less auto-generated expected values.
Should be easier to read and less error prone.

I think that when writing tests is best to be more verbose to make sure
the test is correct and easy to review.

And I think that this can help when you are looking for a big at 2AM, after 10 hours at work

```suggestion
        environ = initializeEnvironment(4, os.getpid())
        environ['LISTEN_FDS'] = 4
        self.patch(os, "environ", )
        sddaemon = ListenFDs.fromEnvironment()
        # The first 3 are standard streams.
        # Our extra streams are added after them.
        self.assertEqual([3, 4, 5, 6], sddaemon.inheritedDescriptors())
```